homeassistant/components/sonos/alarms.py

```diff
@@ -23,48 +23,72 @@ class SonosAlarms(SonosHouseholdCoordinator):
     def __init__(self, *args: Any) -> None:
         """Initialize the data."""
         super().__init__(*args)
-        self._alarms: dict[str, Alarm] = {}
+        self.alarms: Alarms = Alarms()
+        self.created_alarm_ids: set[str] = set()
 
     def __iter__(self) -> Iterator:
         """Return an iterator for the known alarms."""
-        alarms = list(self._alarms.values())
-        return iter(alarms)
+        return iter(self.alarms)
 
     def get(self, alarm_id: str) -> Alarm | None:
         """Get an Alarm instance."""
-        return self._alarms.get(alarm_id)
+        return self.alarms.get(alarm_id)
 
-    async def async_update_entities(self, soco: SoCo) -> bool:
+    async def async_update_entities(
+        self, soco: SoCo, update_id: int | None = None
+    ) -> None:
         """Create and update alarms entities, return success."""
-        try:
-            new_alarms = await self.hass.async_add_executor_job(self.update_cache, soco)
-        except (OSError, SoCoException) as err:
-            _LOGGER.error("Could not refresh alarms using %s: %s", soco, err)
-            return False
-
-        for alarm in new_alarms:
-            speaker = self.hass.data[DATA_SONOS].discovered[alarm.zone.uid]
-            async_dispatcher_send(
-                self.hass, SONOS_CREATE_ALARM, speaker, [alarm.alarm_id]
-            )
+        updated = await self.hass.async_add_executor_job(
+            self.update_cache, soco, update_id
+        )
+        if not updated:
+            return
+
+        for alarm_id, alarm in self.alarms.alarms.items():
+            if alarm_id in self.created_alarm_ids:
+                continue
+            speaker = self.hass.data[DATA_SONOS].discovered.get(alarm.zone.uid)
```

Comment:
```
alarms.py:            speaker = self.hass.data[DATA_SONOS].discovered.get(alarm.zone.uid)
household_coordinator.py:        discovered = self.hass.data[DATA_SONOS].discovered
speaker.py:        missing_zone = self.hass.data[DATA_SONOS].discovered[uid].zone_name
speaker.py:                speaker = self.hass.data[DATA_SONOS].discovered.get(uid)
speaker.py:                slave = self.hass.data[DATA_SONOS].discovered.get(slave_uid)
speaker.py:        speakers = self.hass.data[DATA_SONOS].discovered.values()
switch.py:            self.speaker = self.hass.data[DATA_SONOS].discovered.get(
```

In a future PR we should hide this implementation detail from the class

Reply:
Yes, absolutely. Need to think about how to best clean this up.